src/BatteryStateOfChargeService.cpp

```diff
@@ -43,21 +48,55 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    if (totalAH_>BATTERY_AMP_HOUR_CAPACITY)
+        return 0;
+    else if (totalAH_<0)
+        return BATTERY_AMP_HOUR_CAPACITY;
+    else
+        return BATTERY_AMP_HOUR_CAPACITY - totalAH_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    return isCharging_;
 }
 
+
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    QTime null_time(0,0,0,0);
+    if (return_time.isNull())
+        return null_time;
+    else
+        return return_time;
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // Update your variables here.
+    if (previousTime_.isNull())
+    {
+        totalAH_=BATTERY_AMP_HOUR_CAPACITY*initialStateOfChargePercent_/100;
+        previousTime_=batteryData.time;
+    }
+    zero_time.setHMS(0,0,0,0);
+    deltaTime_=previousTime_.msecsTo(batteryData.time)/HOUR_TO_SEC;
+    deltaCurrent_=(batteryData.current+previousCurrent_)/2;
+    amph_=deltaTime_*deltaCurrent_;
+    totalAH_+=amph_;
+
+    if (batteryData.current < 0){
+        isCharging_=true;
+        deltaAH_=-1*(BATTERY_AMP_HOUR_CAPACITY-totalAH_);
+
+    }
+    else{
+        isCharging_=false;
+        deltaAH_=totalAH_;
+    }
+    deltaAH_ /= batteryData.current;
```

Comment:
Keep this `deltaAH_` private member I'd say is good, but it's a little confusing that you're calculating it in a few different places. You are already calculating it in to `amph_` on line 84, so I'd suggest just using that calculation instead. Let me know if I'm misunderstanding something.